### PR TITLE
[Function] Integrate package management service to pulsar functionWorker

### LINKED ISSUE #14224: Integrate package service with function
**Is your enhancement request related to a problem? Please describe.**
Currently, I'm implementing PIP-117 https://github.com/apache/pulsar/issues/13302 , changing from using DistributedLog to using local filesystem to store the jars directly in the data folder instead of uploading them into BK. 

However, I found the current function creation will directly store the jar file into DistributedLog by calling `WorkerUtils.uploadFileToBookkeeper` no matter we create function by jar file or functionPkgUrl. It doesn't store the jar file by the unified package service interface. Thus it is hard to store the jars into filesystem by configuration.

**Describe the solution you'd like**
We'd better integrate the package service into function, and use the unified package service api to read and write jar files.

### PR DESCRIPTION
**Pull request by gaozhangmin**
Fixes #14224
### Motivation

The current function creation will directly store the jar file into DistributedLog by calling WorkerUtils.uploadFileToBookkeeper no matter we create function by jar file or functionPkgUrl. It doesn't store the jar file by the unified package service interface

### Modifications
1、Add a new configuration `functionsWorkerEnablePackageManagement `, whether work uses packageManagement service to store packages.
2、If We enable `functionsWorkerEnablePackageManagement`, while creating function/sink/source, the package will be uploaded to packageManagement store.

### Documentation

Check the box below or label this PR directly (if you have committer privilege).

Need to update docs? 

- [ ] `doc-required` 
  
  (If you need help on updating docs, create a doc issue)
  
- [x] `no-need-doc` 
  
  (Please explain why)
  
- [ ] `doc` 
  
  (If this PR contains doc changes)

### CODE REVIEW

File: conf/functions_worker.yml [hunk 1]
```diff
@@ -26,6 +26,13 @@ workerHostname: localhost
 workerPort: 6750
 workerPortTls: 6751
 
+# The metadata store URL
+# Examples:
+# * zk:my-zk-1:2181,my-zk-2:2181,my-zk-3:2181
+# * my-zk-1:2181,my-zk-2:2181,my-zk-3:2181 (will default to ZooKeeper when the schema is not specified)
+# * zk:my-zk-1:2181,my-zk-2:2181,my-zk-3:2181/my-chroot-path (to add a ZK chroot path)
+metadataStoreUrl: zk:localhost:2181
```
Comment (nlu90) [inline conf/functions_worker.yml:34]: what's the difference with the following `configurationMetadataStoreUrl`?

File: conf/functions_worker.yml [hunk 2]
```diff
@@ -45,6 +52,10 @@ metadataStoreCacheExpirySeconds: 300
 ################################
 # Function package management
 ################################
+packagesManagementStorageProvider:
+    org.apache.pulsar.packages.management.storage.bookkeeper.BookKeeperPackagesStorageProvider
+packagesManagementLedgerRootPath: /ledgers
+packagesReplicas: 1
```
Comment (nlu90) [inline conf/functions_worker.yml:58]: `broker.conf` and `standalone.conf` also have these same configurations.

If you added them here, what's the relationship between these duplicated configuration?

Do you want to remove them from other conf files or they will have some override relationship?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java [hunk 3]
```diff
@@ -780,14 +780,14 @@ public void start() throws PulsarServerException {
 
             acquireSLANamespace();
 
-            // start function worker service if necessary
-            this.startWorkerService(brokerService.getAuthenticationService(), brokerService.getAuthorizationService());
-
             // start packages management service if necessary
             if (config.isEnablePackagesManagement()) {
                 this.startPackagesManagementService();
             }
```
Comment (nlu90) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java:786]: If you make function worker depend on package management service, then shouldn't the package management service be always enabled when function worker is enabled?

Currently it seems they are separately configured, which may result the case that function worker is enabled but package management service is not running.

**Review by nlu90 — CHANGES_REQUESTED** [review]

File: conf/functions_worker.yml (hunk 1 above)
Comment (gaozhangmin) [reply conf/functions_worker.yml:34]: metadataStoreUrl is used for  BookKeeperPackagesStorage#initializeDlogNamespace。If we want to start  BookKeeperPackagesStorage storage service  in a separately running workers. we must support this configuration.

File: conf/functions_worker.yml (hunk 2 above)
Comment (gaozhangmin) [reply conf/functions_worker.yml:58]: Run Functions-worker separately, independent with broker, in this situation, if we want to start packageManagement service in worker, We should have this configurations.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java (hunk 3 above)
Comment (gaozhangmin) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java:786]: There are two situations:
1、Run Functions-worker with brokers. If broker enables  function worker without enabling packageManagement,  packageManagement service is enabled in worker by default. Else, worker will use packageManagement service enabled by broker.
2、Run Functions-worker separately. Without broker, We enable packageManagement service  in worker by default

File: conf/functions_worker.yml (hunk 1 above)
Comment (gaozhangmin) [reply conf/functions_worker.yml:34]: We can be back to use configurationMetadataStoreUrl if metadataStoreUrl is not set.

Comment (gaozhangmin) [reply conf/functions_worker.yml:34]: @nlu90

**Review by freeznet — CHANGES_REQUESTED** [review]
Thanks for the contribution, i think this PR will bring some breaking changes: 
- removes the BK package storage for future versions
- make package management service as a internal service for functions worker

and the above breaking changes may cause backward compatibility issue. We need consider this carefully, and it is better to send out a discussion thread on the mail-list first.

Also, here is my thought about https://github.com/apache/pulsar/issues/14224, and there are two tasks:
1. when create / update functions with user provide `packageUrl` from package management service
  - validate the pacakge existance
  - skip `WorkerUtils.uploadFileToBookkeeper` since this procudure is redundant
  - allow download from packageURL when start functions in different runtimes
2. use package management service to host user upload function package
  - the package management service should be an external service for functions worker (the package management service is disabled by default for current releases), and all functionalities should enabled only package management service is enabled.
  - use pulsar-admin apis to call package service APIs from functions worker
  - keep consistancies between two service (function worker service & package management service), such as user can  delete a package which is already used by other functions, and it will breaks all other functions. On the other hand, when user delete a function, it should delete the package from package management service if the package is refer by the delete function only. 
  - functions worker should be able to config if use package management service to host the package.

**Comment by gaozhangmin** [comment]
@freeznet  For task one, Now, user can create or update a function/sink/source by: 
1、local path to the JAR file for the function (if the function is written in Java). 
2、`functionPkgUrl`  which starts with http/https/file/package URL from packages management service

I think if We config function worker to use packageManagement service, We should upload package to it either we create functions by local path or `functionPkgUrl with http/file/https`.

**Comment by gaozhangmin** [comment]
For backward compatibility, I want to make this changes. @freeznet 
1、We only upload and download package from packageManagement service If We set `enablePackagesManagement=true` in broker and add a new configuration to function worker to set if we use packageManagement service to host package or not.
2、Remove the internal packageManagement service in function worker.
Thanks again for your reviews.

**Comment by freeznet** [comment]
@gaozhangmin 

> 1、We only upload and download package from packageManagement service If We set `enablePackagesManagement=true` in broker and add a new configuration to function worker to set if we use packageManagement service to host package or not.
> 2、Remove the internal packageManagement service in function worker.

Thanks, lgtm

**Comment by gaozhangmin** [comment]
@freeznet  PTAL. Unit test will be added lately.

**Comment by liangyuanpeng** [comment]
/ping @freeznet  @hangc0276  @zymap  Let me push it and please review this PR, Thanks.

**Comment by nlu90** [comment]
@freeznet @zymap @hangc0276 Can you guys take a look again?

**Review by merlimat — APPROVED** [review]

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionActioner.java [hunk 4]
```diff
@@ -223,6 +222,13 @@ private void downloadFile(File pkgFile, boolean isPkgUrlProvided, FunctionMetaDa
 
         if (downloadFromHttp) {
             FunctionCommon.downloadFromHttpUrl(pkgLocationPath, tempPkgFile);
+        } else if (workerConfig.isFunctionsWorkerEnablePackageManagement()) {
```
Comment (zymap) [inline pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionActioner.java:225]: We download with the package service at line 232, why need a new branch? And if we download failed, don't we need to throw the exception?

**Review by zymap — APPROVED** [review]
LGTM

**Review by freeznet — APPROVED** [review]

**Pull request merged**